selfservice/strategy/code/strategy_registration.go

```diff
@@ -0,0 +1,267 @@
+	"github.com/ory/kratos/ui/container"
+	"github.com/ory/kratos/ui/node"
+	"github.com/ory/kratos/x"
+	"github.com/ory/x/urlx"
+)
+
+var _ registration.Strategy = new(Strategy)
+
+// Update Registration Flow with Code Method
+//
+// swagger:model updateRegistrationFlowWithCodeMethod
+type updateRegistrationFlowWithCodeMethod struct {
+	// The identity's traits
+	//
+	// required: true
+	Traits json.RawMessage `json:"traits" form:"traits"`
+
+	// The OTP Code sent to the user
+	//
+	// required: false
+	Code string `json:"code" form:"code"`
+
+	// The CSRF Token
+	CSRFToken string `json:"csrf_token" form:"csrf_token"`
+
+	// Method to use
+	//
+	// This field must be set to `code` when using the code method.
+	//
+	// required: true
+	Method string `json:"method" form:"method"`
+
+	// Transient data to pass along to any webhooks
+	//
+	// required: false
+	TransientPayload json.RawMessage `json:"transient_payload,omitempty" form:"transient_payload"`
+
+	// Resend restarts the flow with a new code
+	//
+	// required: false
+	Resend string `json:"resend" form:"resend"`
+}
+
+func (s *Strategy) RegisterRegistrationRoutes(*x.RouterPublic) {}
+
+func (s *Strategy) HandleRegistrationError(w http.ResponseWriter, r *http.Request, flow *registration.Flow, body *updateRegistrationFlowWithCodeMethod, err error) error {
+	if flow != nil {
+		if body != nil {
+			action := flow.AppendTo(urlx.AppendPaths(s.deps.Config().SelfPublicURL(r.Context()), registration.RouteSubmitFlow)).String()
+			for _, n := range container.NewFromJSON(action, node.CodeGroup, body.Traits, "traits").Nodes {
+				// we only set the value and not the whole field because we want to keep types from the initial form generation
+				flow.UI.Nodes.SetValueAttribute(n.ID(), n.Attributes.GetValue())
+			}
+		}
+
+		flow.UI.SetCSRF(s.deps.GenerateCSRFToken(r))
+	}
+
+	return err
+}
+
+func (s *Strategy) PopulateRegistrationMethod(r *http.Request, rf *registration.Flow) error {
+	return s.PopulateMethod(r, rf)
+}
+
+type options func(*identity.Identity) error
+
+func WithCredentials(via identity.CodeAddressType, usedAt sql.NullTime) options {
+	return func(i *identity.Identity) error {
+		return i.SetCredentialsWithConfig(identity.CredentialsTypeCodeAuth, identity.Credentials{Type: identity.CredentialsTypePassword, Identifiers: []string{}}, &identity.CredentialsCode{AddressType: via, UsedAt: usedAt})
+	}
+}
+
+func (s *Strategy) handleIdentityTraits(ctx context.Context, f *registration.Flow, traits json.RawMessage, transientPayload json.RawMessage, i *identity.Identity, opts ...options) error {
+	f.TransientPayload = transientPayload
+	if len(traits) == 0 {
+		traits = json.RawMessage("{}")
+	}
+
+	// we explicitly set the Code credentials type
+	i.Traits = identity.Traits(traits)
+	if err := i.SetCredentialsWithConfig(s.ID(), identity.Credentials{Type: s.ID(), Identifiers: []string{}}, &identity.CredentialsCode{UsedAt: sql.NullTime{}}); err != nil {
```

Comment:
Will identifiers be hydrated at a later stage? If not, login will probably fail because the identifier is missing.